### ISSUE TITLE
Change Serilog for ZeroLog for NET

### ISSUE THREAD

**Issue by cdhanna** [beamable/BeamableProduct#3667]
Serilog is lacking, 
1. not performant (allocates a lot)
2. hard to customize (they made everything private)
3. we aren't using serilog to its designed use case; instead we are using the bare minimum level of features...

We don't _need_ serilog to meet our logging needs, and its a pain in the tookus. We should change it for a better solution, perhaps like ZeroLog. We should investigate a solution that is easy to customize, fast, and not a pain.

**Comment by Leinnan** [comment]
Other options- 
https://github.com/Cysharp/ZLogger
https://github.com/NLog/NLog

**Comment by Leinnan** [comment]
## ZeroLog

- API is more different from our existing one in comparison to other options
- fast
- structured logging support
## ZLogger

Seems to be the most promising one
- change log level at runtime could be handled like here: [tests/ZLogger.Tests/BuilderTest.cs](https://github.com/Cysharp/ZLogger/blob/0b80aeb9d06e2b021469babbc932ccbe8df5507c/tests/ZLogger.Tests/BuilderTest.cs#L13)
- structured log format support
- support for adding log properties in scopes:
```csharp
	using( CliSerilogProvider.LogContext.Value.BeginScope(new Dictionary<string,object>  
      {  
          {"AppContext", args.AppContext},  
          {"Command", command}  
      })){  
BeamableLogger.LogVerbose("Starting command");
BeamableLogger.LogVerbose("Running command");}
```


## NLog

NLog doesn't support AOT (yet! It is planned for 6.0).
Also this solution doesn't seem to be as much conserned about allocations as other possible solutions.

**Comment by Leinnan** [comment]
I did some tests for Zlogger on `zlogger-tests` branch if anyone is interested.

**Comment by Leinnan** [comment]
ZeroLog according to it's author is faster than ZLogger when using default API, but slower than using ZLogger source generated one: https://x.com/Lucas_Trz/status/1731774659350126873

Both of them being faster than Serilog and NLog: 
![image](https://github.com/user-attachments/assets/182f173d-79cf-4207-b2a4-4eb6974e28e6)

**Comment by cdhanna** [comment]
So it sounds like it is between ZeroLog and ZLogger. 

1. do they both support multiple sinks?
2. would they be able to integrate with ansi-console, like our serilog sink is doing? (essentially, there is a serilog sink that outputs to ansi-console)
3. can we inject code-in-the-middle to do masking support ourselves? 

At the moment, what is your recommendation, and why?

**Comment by Leinnan** [comment]
1. Yes, they do.
2. I think both of them can. I test it out for ZLogger and in order to get first version it took me few lines: 
```csharp
logging.AddZLoggerInMemory(processor =>
{
	processor.MessageReceived += renderedLogString => 
	{
		AnsiConsole.WriteLine(renderedLogString);
	};
});
```
For ZeroLog it would require to implement a `Appender` but it is doable.
3. This one is tricky. I didn't found how to do it for ZeroLog. Maybe it could be done at Appender level. ZLogger uses Microsoft builder for logging which offers some extensions for masking, but I did not get it to work with ZLogger, but I did not yet explore it too much.

I would go for ZLogger:
- they are providing "default" Microsoft API for logging with opt in for their methods with more performance wins, ZeroLog has its own API
- based on code samples transition should be easier to ZLogger
- while ZeroLog is faster for default case, we could just `optimize`  the most used cases by providing source generated logs for them only and that should result in similar, if not better performance with slighty bigger memory usage compared to ZeroLog(but still way smaller than Serilog/NLog)

**Comment by cdhanna** [comment]
Cool; yeah I'm curious to know more about the source-gen logs. Sounds like the next step (like you said in stand-up) is to continue the research with some prototyping and benchmarking, in an explorative capacity.

**Comment by Leinnan** [comment]
Yeah, writing them should be simple, and I am also curious about possible performance wins:
```csharp
public static partial class MyLogger
{
    [ZLoggerMessage(LogLevel.Information, "Bar: {x} {y}")]
    public static partial void Bar(this ILogger<Foo> logger, int x, int y);
}
```
> This can achieve the highest performance. It's also possible to use special format specifiers like :json.
